crates/turborepo-repository/src/change_mapper.rs

```diff
@@ -0,0 +1,233 @@
+//! Maps changed files to changed packages in a repository.
+//! Used for both `--filter` and for isolated builds.
+
+use std::collections::HashSet;
+
+use turbopath::{AbsoluteSystemPath, AnchoredSystemPath, AnchoredSystemPathBuf};
+use wax::Program;
+
+use crate::package_graph::{ChangedPackagesError, PackageGraph, WorkspaceName};
+
+// We may not be able to load the lockfile contents, but we
+// still want to be able to express a generic change.
+pub enum LockfileChange {
+    Empty,
+    WithContent(Vec<u8>),
+}
+
+pub enum PackageChanges {
+    All,
+    Some(HashSet<WorkspaceName>),
+}
+
+pub struct ChangeMapper<'a> {
+    pkg_graph: &'a PackageGraph,
+
+    global_deps: Vec<String>,
+    ignore_patterns: Vec<String>,
+}
+
+impl<'a> ChangeMapper<'a> {
+    const DEFAULT_GLOBAL_DEPS: [&'static str; 2] = ["package.json", "turbo.json"];
+
+    pub fn new(
+        pkg_graph: &'a PackageGraph,
+        global_deps: Vec<String>,
+        ignore_patterns: Vec<String>,
+    ) -> Self {
+        Self {
+            pkg_graph,
+            global_deps,
+            ignore_patterns,
+        }
+    }
+
+    pub fn changed_packages(
+        &self,
+        changed_files: HashSet<AnchoredSystemPathBuf>,
+        lockfile_change: Option<LockfileChange>,
+    ) -> Result<PackageChanges, ChangeMapError> {
+        let global_change =
+            self.repo_global_file_has_changed(&Self::DEFAULT_GLOBAL_DEPS, &changed_files)?;
+
+        if global_change {
+            return Ok(PackageChanges::All);
+        }
+
+        // get filtered files and add the packages that contain them
+        let filtered_changed_files = self.filter_ignored_files(changed_files.iter())?;
+        let mut changed_pkgs =
+            self.get_changed_packages(filtered_changed_files.into_iter(), self.pkg_graph)?;
+
+        match lockfile_change {
+            Some(LockfileChange::WithContent(content)) => {
+                // if we run into issues, don't error, just assume all packages have changed
+                let Ok(lockfile_changes) = self.get_changed_packages_from_lockfile(content) else {
+                    return Ok(PackageChanges::All);
+                };
+
+                changed_pkgs.extend(lockfile_changes);
+
+                Ok(PackageChanges::Some(changed_pkgs))
+            }
+            // We don't have the actual contents, so just invalidate everything
+            Some(LockfileChange::Empty) => Ok(PackageChanges::All),
+            None => Ok(PackageChanges::Some(changed_pkgs)),
+        }
+    }
+
+    fn repo_global_file_has_changed(
+        &self,
+        default_global_deps: &[&str],
+        changed_files: &HashSet<AnchoredSystemPathBuf>,
+    ) -> Result<bool, turborepo_scm::Error> {
+        let global_deps = self.global_deps.iter().map(|s| s.as_str());
+        let filters = global_deps.chain(default_global_deps.iter().copied());
+        let matcher = wax::any(filters).unwrap();
```

Comment:
I'm pretty sure tracing through the code that these globs haven't been validated. We should probably throw here since it's user input.